Sources/Sentry/SentryScopeSyncC.c

```diff
@@ -0,0 +1,225 @@
+#include <pthread.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#define NUMBER_OF_FIELDS 9
+
+static const char *userJSON;
+static const char *distJSON;
+static const char *contextJSON;
+static const char *environmentJSON;
+static const char *tagsJSON;
+static const char *extrasJSON;
+static const char *fingerprintJSON;
+static const char *levelJSON;
+
+static long maxCrumbs = 0;
+static int currentCrumb = 0;
+static const char **breadcrumbs; // dynamic array of char arrays
+static const char *breadcrumbsStart = "\"breadcrumbs\":[";
+
+static void
+add(char *destination, const char *source)
+{
+    if (source != NULL) {
+        strcat(destination, source);
+        strcat(destination, ",");
+    }
+}
+
+static size_t
+getSize(const char *str)
+{
+    size_t size = 0;
+    if (str != NULL) {
+        size = strlen(str);
+    }
+    return size;
+}
+
+static size_t
+getRawBreadcrumbSize(void)
+{
+    size_t addionitalChars = 3;
+    return sizeof(breadcrumbsStart) + maxCrumbs + addionitalChars + 1;
+}
+
+static size_t
+getBreadcrumbSize(void)
+{
+    size_t size = getRawBreadcrumbSize();
+    for (int i = 0; i < maxCrumbs; i++) {
+        size += getSize(breadcrumbs[i]);
+    }
+    return size;
+}
+
+static void
+addBreadcrumbs(char *destination)
+{
+    size_t size = getBreadcrumbSize();
+    char *crumbs = malloc(size);
+
+    // No crumbs nothing to add
+    if (size == getRawBreadcrumbSize()) {
+        return;
+    }
+
+    strcat(crumbs, breadcrumbsStart);
+    for (int i = 0; i < maxCrumbs; i++) {
+        if (breadcrumbs[i] != NULL) {
+            strcat(crumbs, "{");
+            strcat(crumbs, breadcrumbs[i]);
+            strcat(crumbs, "},");
+        }
+    }
+
+    size_t length = strlen(crumbs);
+    crumbs[length - 1] = ']';
+    crumbs[length] = '\0';
+    strcat(crumbs, ",");
+
+    strcat(destination, crumbs);
+
+    free(crumbs);
+}
+
+/**
+ * We don't lock access to the properties in this method as this is called when a crash occurs.
```

Comment:
Well, not quite.
Your `set` function actually calls `free` before resetting the pointer. Which means that if you crash in between, your crash handler will have a use-after-free.